compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategiesGPU.cpp

```diff
@@ -266,18 +295,141 @@ void ReductionStrategyStagedThreadDistribution::compute(
     else
       vectorSizeStage2 = 1;
   }
+
+  // Stage 3
+  // -------
+  // Stage 3 is non-ambiguous and does not ned to be configured (always uses
+  // warp shuffles).
+}
+
+/// Encodes a strategy targeted at (very) small reductions, for which other
+/// strategies perform poorly.
+///
+/// In the case of small reductions, we cannot make an efficient use of warp
+/// shuffles. Instead, try to take advantage of caches.
+/// This strategy aims at running the reduction sequentially within each thread
+/// and taking parallelism from outer dimensions that we would otherwise use
+/// for block-level parallelism.
+///
+/// There are 2 cases:
+///   1. we can find good divisors of outer parallel dimensions and avoid
+///      creating dynamic tile sizes. We can then vectorize to the reduction
+///      size.
+///   2. we cannot find good divisors, we pay the price of dynamic loops.
+///
+// TODO: Refine 1. with linalg splitting on the reduction dimension.
+// TODO: Refine 2. with linalg splitting on the parallel dimension.
+//
+// Note: All this is to be able to handle very small and small-ish reductions
+// without catastrophic regressions.
+// TODO: Add a strategy based on segmented scans, which can allow us to force
+// sizes that don't divide properly into warp shuffles.
+class SmallReductionStrategy : public GPUReductionStrategy {
+ public:
+  /// `hasTrailingElementwise` is currently used to guard against pathological
+  /// cases where IREE can't bound a buffer and crashes.
+  // TODO: Fix codegen/Common/PadDynamicAlloc.cpp which calls into upstream
```

Comment:
Is there an issue for this?

Reply:
I maintain a stack of issues in a doc atm, I'll make sure to turn into a github issue.